mcpserver/mcpserver.go

```diff
@@ -0,0 +1,107 @@
+package mcpserver
+
+import (
+	"context"
+
+	"dagger.io/dagger"
+	"github.com/dagger/container-use/environment"
+	"github.com/dagger/container-use/repository"
+	"github.com/mark3labs/mcp-go/mcp"
+)
+
+func dagFromContext(ctx context.Context) *dagger.Client {
+	dag, ok := ctx.Value(daggerClientKey{}).(*dagger.Client)
+	if !ok {
+		panic("dagger client not found in context")
+	}
+	return dag
+}
+
+type Request interface {
+	isRequest()
+}
+
+type BaseRequest struct {
+	Explanation string `json:"explanation"`
+}
+
+func (BaseRequest) isRequest() {}
+
+type BaseRepositoryRequest struct {
+	BaseRequest
+
+	EnvironmentSource string `json:"environment_source"`
+}
+
+type BaseEnvironmentRequest struct {
+	BaseRepositoryRequest
+
+	EnvironmentID string `json:"environment_id"`
+}
+
+type Response any
+
+type ToolResponse[T Response] struct {
+	Message string
+	Data    T
+}
+
+func openRepositoryFromRequest(ctx context.Context, request BaseRepositoryRequest) (*repository.Repository, error) {
+	repo, err := repository.Open(ctx, request.EnvironmentSource)
+	if err != nil {
+		return nil, err
+	}
+	return repo, nil
+}
+
+func openEnvironmentFromRequest(ctx context.Context, request BaseEnvironmentRequest) (*repository.Repository, *environment.Environment, error) {
+	repo, err := openRepositoryFromRequest(ctx, request.BaseRepositoryRequest)
+	if err != nil {
+		return nil, nil, err
+	}
+	env, err := repo.Get(ctx, dagFromContext(ctx), request.EnvironmentID)
+	if err != nil {
+		return nil, nil, err
+	}
+	return repo, env, nil
+}
+
+func parseBaseRequest(request mcp.CallToolRequest) (BaseRequest, error) {
+	explanation, err := request.RequireString("explanation")
```

Comment:
🐛 **Bug (High):** Inconsistency: explanation field is not marked as required in args.go but parseBaseRequest uses RequireString. Should use GetString with empty default to match existing behavior in other tools.

```suggestion
explanation := request.GetString("explanation", "")
```